### DISCUSSION TITLE
Problems displaying very large polygons

### DISCUSSION THREAD

**Discussion by JosephTRichardson** [pelicanmapping/osgearth#2294] [Support Q&A]
Hi, we're developing a plugin for an osgEarth/SIMDIS SDK-based application. Our plugin is supposed to help users manage large geospatial datasets and pick out areas of interest to export. So, for example, they might want to look at a global dataset like DTED0. To style that area of coverage, we are merging together the coverage area of the dataset into a geometry and creating a polygon of it as an osgEarth Feature. But, two problems: one, the application gets really, really slow -- I'm not sure which layer of the application is having trouble handling that, whether osgEarth or something higher up. And two, it looks very weird and bad. We've tried several different methods of styling them, extrusion at various altitudes and draping -- and it just looks wrong.

With extrusion, the plane of the polygon seems to get stuck in the plane of the earth, and whole large portions of polygon disappear behind the land. It's not behind the land imagery layer: even if we remove the .earth file and have a blank earth, the polygon is lost somewhere underground. If we set it really high, that helps some, but creates its own problems. Examples:

>     osgEarth::Style polyStyle;
>     polyStyle.getOrCreate<osgEarth::PolygonSymbol>()->fill()->color()
>         = osgEarth::Color(osgEarth::Color::Maroon, 1.0f);
>     polyStyle.getOrCreate<osgEarth::ExtrusionSymbol>()->height() =
>         250000.0; // meters MSL

![Screenshot 2023-06-01 141058](https://github.com/gwaldron/osgearth/assets/7881972/0089e962-6646-4b09-a44a-df32dcd16628)
![Screenshot 2023-06-01 141139](https://github.com/gwaldron/osgearth/assets/7881972/1154608a-5db0-4f0f-916a-78b8d0f6b726)

With draping, it looks almost passably decent in North America, but creates strange artifacts. For the whole eastern hemisphere, it seems to just blow up. DTED0 doesn't cover the ocean and is not supposed to look like this.

>     osgEarth::Style polyStyle;
>     polyStyle.getOrCreate<osgEarth::PolygonSymbol>()->fill()->color()
>         = osgEarth::Color(osgEarth::Color::Maroon, 1.0f);
>     polyStyle.getOrCreate<osgEarth::AltitudeSymbol>()->clamping() =
>         osgEarth::AltitudeSymbol::CLAMP_TO_TERRAIN;
>     polyStyle.getOrCreate<osgEarth::AltitudeSymbol>()->technique() =
>         osgEarth::AltitudeSymbol::TECHNIQUE_DRAPE;

![Screenshot 2023-06-01 143335](https://github.com/gwaldron/osgearth/assets/7881972/60d0fa73-f92c-45f9-9d98-b431103ce102)
![Screenshot 2023-06-01 143404](https://github.com/gwaldron/osgearth/assets/7881972/8e82e077-f752-4fcf-8783-ff0df974790d)

Since we're dealing with several layers of application here -- osgEarth, SIMDIS SDK, and another plugin layer built on top of that -- it's possible that one of the other layers is causing issues. But it appears likely to me to be an issue with osgEarth styling. Do you have any advice on what to do about this, either how to improve performance or how to style it better?

**Comment by gwaldron** [comment]
Are you able to provide the shapefile (shp/shx/dbf)?

**Reply by JosephTRichardson** [reply]
Hi, thanks for the reply. We weren't exporting a shapefile, but I made one for you.

[dted0.geojson.zip](https://github.com/gwaldron/osgearth/files/11637196/dted0.geojson.zip)
[dted0.shp.zip](https://github.com/gwaldron/osgearth/files/11637191/dted0.shp.zip)
(The GeoJSON is directly exported from our program. The shapefile -- it was surprisingly difficult to get a shapefile out of GDAL, so that is exported out of QGIS from the GeoJSON.)

In our code, we are going directly from geometry to feature:

```
 	std::shared_ptr<OGRGeometry> geometry = rasterFile->getCoverageGeometry();
          // getCoverageGeometry() creates a union of the bounding boxes of all tiles
 	OGRGeometryH geometryHandle = geometry->ToHandle(geometry.get());
 	osgEarth::Symbology::Geometry* osgGeo = osgEarth::Util::OgrUtils::createGeometry(geometryHandle);
 	osgEarth::Feature* polyFeature = new osgEarth::Feature(osgGeo, geoSRS);
 	m_pFeatureNode = new osgEarth::FeatureNode(polyFeature, polyStyle, options);
```

The polygon renders without major issues in QGIS and Google Earth. In osgEarth, it appears it's having problems with the vertices of internal rings not connecting. They APPEAR to be connected correctly in our geometry/file (I ran `OGRGeometry->closeRings()` before getting the geometry from `getCoverageGeometry()`). We have made some progress today. My colleague says the latest attempt at draping the polygon has improved the performance a great deal -- everything is smooth again. And if we turn `LineSymbol->stroke()` to 0 width, it _looks_ a good bit better.

![Screenshot 2023-06-02 110141](https://github.com/gwaldron/osgearth/assets/7881972/cbc41468-a45c-4aff-aabf-b1d6168bd1e9)
That missing strip through Africa is supposed to be there.

![Screenshot 2023-06-02 110219](https://github.com/gwaldron/osgearth/assets/7881972/6ec56003-d028-4a6d-9ba0-a284fd63a33c)
Those artifacts over North America are still there, apparently the result of some internal ring that didn't fully close.

Here's an even more obvious example of rings not fully closing:

![today](https://github.com/gwaldron/osgearth/assets/7881972/7bee64bb-22cd-4159-8448-8e1f5f8cd75f)

**Reply by JosephTRichardson** [reply]
My colleague just showed me this. If you have JUST stroke, no fill, the outline of the polygon appears to be okay. If you add fill, that's where issues occur. (This is not our code but the plugin layer code, but under the hood it's just osgEarth styling.)

![stroke, no fill](https://github.com/gwaldron/osgearth/assets/7881972/de92f8df-cafe-4b70-a54a-2eadded00996)

![fill](https://github.com/gwaldron/osgearth/assets/7881972/4ba0c073-151f-4389-814d-4488386ef907)
